xlators/protocol/server/src/server-handshake.c

```diff
@@ -45,20 +45,76 @@ server_getspec(rpcsvc_request_t *req)
     gf_getspec_rsp rsp = {
         0,
     };
+    struct stat stbuf = {
+        0,
+    };
+    char volpath[PATH_MAX] = {
+        0,
+    };
+    int32_t spec_fd = -1;
+    xlator_t *this = req->svc->xl;
+    server_conf_t *conf = this->private;
 
     ret = xdr_to_generic(req->msg[0], &args, (xdrproc_t)xdr_gf_getspec_req);
     if (ret < 0) {
         // failed to decode msg;
         req->rpc_err = GARBAGE_ARGS;
         op_errno = EINVAL;
-        goto fail;
+        rsp.spec = "<this method is not in use, use glusterd for getspec>";
+        rsp.op_errno = gf_errno_to_error(op_errno);
+        rsp.op_ret = -1;
+        goto out;
     }
 
-    op_errno = ENOSYS;
-fail:
-    rsp.spec = "<this method is not in use, use glusterd for getspec>";
-    rsp.op_errno = gf_errno_to_error(op_errno);
-    rsp.op_ret = -1;
+    char *volid = args.key;
+    ret = snprintf(volpath, PATH_MAX - 1, "%s/%s.vol", conf->volfile_dir,
+                   volid);
+    if (ret == -1) {
+        gf_msg(this->name, GF_LOG_ERROR, errno, 0, "failed to copy volfile");
+        goto out;
+    }
+
+    ret = sys_stat(volpath, &stbuf);
+    if (ret < 0) {
+        goto out;
+    }
+
+    spec_fd = open(volpath, O_RDONLY);
+    if (spec_fd < 0) {
+        gf_msg("glusterd", GF_LOG_ERROR, errno, 0, "Unable to open %s (%s)",
+               volpath, strerror(errno));
+        goto out;
+    }
+    ret = stbuf.st_size;
+
+    if (ret > 0) {
+        rsp.spec = CALLOC(ret + 1, sizeof(char));
+        if (!rsp.spec) {
+            gf_msg(this->name, GF_LOG_ERROR, errno, 0, "no memory");
+            ret = -1;
+            op_errno = ENOMEM;
+            goto out;
+        }
+        ret = sys_read(spec_fd, rsp.spec, ret);
+    }
+
+out:
+    if (spec_fd >= 0)
+        sys_close(spec_fd);
+
+    rsp.op_ret = ret;
+    if (rsp.op_ret < 0) {
+        if (!op_errno)
+            op_errno = ENOENT;
```

Comment:
Wouldn't it be better to set op_errno in all error paths to the appropriate value instead of returning a possibly misleading error ?

Reply:
Ack, handled now